include/deal.II/distributed/cell_weights.h

```diff
@@ -0,0 +1,170 @@
+     * objects, we can take those into account by introducing individual weights
+     * for different cells.
+     *
+     * This class allows to use information about the FiniteElement on each cell
+     * to determine the weight through the hp::DoFHandler to the connected
+     * parallel::distributed::Triangulation, in which callback function of this
+     * class will be connected to its corresponding signal.
+     *
+     * An object of this class needs to exist for every DoFHandler associated
+     * with the Triangulation we work on to achieve satisfying work balancing.
+     *
+     * A small code snippet follows explaining how to achieve each cell
+     * being weighted by its current number of degrees of freedom.
+     * @code
+     * parallel::distributed::CellWeights<dim, spacedim>
+     *   cell_weights(hp_dof_handler);
+     * cell_weights.register_ndofs_weighting();
+     * @endcode
+     * The weighting function can be changed anytime. Even more ambitious
+     * approaches are possible by submitting customized functions, e.g.
+     * @code
+     * cell_weights.register_custom_weighting(
+     *   [](const FiniteElement<dim, spacedim> &) -> unsigned int {
+     *   return std::pow(fe.dofs_per_cell, 1.6);
+     * });
+     * @endcode
+     * @note Keep in mind that the return values are restricted to unsigned integers.
+     *
+     * @note Be aware that this class registers the weight function to the
+     * Triangulation's signal during its construction. If the Triangulation
+     * associated with the DoFHandler changes during the lifetime of the
+     * latter, an assertion will be triggered in the weight_callback function.
+     * It is recommended to create a separate object in this case and to destroy
+     * previous one.
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2018
+     */
+    template <int dim, int spacedim = dim>
+    class CellWeights
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param[in] dof The hp::DoFHandler which will be used to
+       *    determine each cell's degrees of freedom.
+       */
+      CellWeights(const dealii::hp::DoFHandler<dim, spacedim> &dof_handler);
+
+      /**
+       * Destructor.
+       */
+      ~CellWeights();
+
+      /**
+       * Choose a constant weight on each cell.
+       */
+      void
+      register_constant_weighting();
+
+      /**
+       * Choose a weight reflecting the number of degrees of freedom on each
+       * cell.
+       */
+      void
+      register_ndofs_weighting();
+
+      /**
+       * Choose a weight reflecting the number of degrees of freedom
+       * <i>squared</i> on each cell.
+       */
+      void
+      register_ndofs_squared_weighting();
+
+      /**
+       * Register a custom cell weight by providing a function as a parameter.
+       */
+      void
+      register_custom_weighting(
+        std::function<unsigned int(const FiniteElement<dim, spacedim> &)>
+          custom_function);
```

Comment:
I think that a better (or additional) interface would take an object that takes a `hp::DoFHandler::active_cell_iterator` rather than a finite element. There are cases where the work is dependent on what you want to *do* on a cell, rather than just the element. I think step-42 is a case in point. In that case, one might want to do
```
  [](const hp::DoFHandler::active_cell_iterator &cell) -> unsigned int
  {
    if (...)
      return cell->get_fe().dofs_per_cell squared;
    else
      return 100 * cell->get_fe().dofs_per_cell squared;  // we do something expensive on this cell
```

Reply:
also, I think you want to take a `const std::function<....> &` reference

Reply:
I agree! I think that we should let our lambda function take two arguments, the cell itself and the active finite element.

Consider the case in 2D where four cells on the boundary get coarsened into one, which will lie on the boundary after the coarsening process. We want to calculate the weight depending on whether a cell is on the boundary, but if we pick one of the cells that is not on the boundary to determine the amount of degrees of freedom on the parent cell, we lose this information. So I guess the combination of both should do the trick.

Reply:
OK.